bascule/key/cache.go

```diff
@@ -137,14 +152,24 @@ func (cache *multiCache) copyPairs() map[string]Pair {
 	return newPairs
 }
 
-func (cache *multiCache) ResolveKey(keyID string) (pair Pair, err error) {
+func (cache *multiCache) ResolveKey(ctx context.Context, keyID string) (pair Pair, err error) {
 	var ok bool
 	pair, ok = cache.fetchPair(keyID)
 	if !ok {
 		cache.update(func() {
 			pair, ok = cache.fetchPair(keyID)
+			if isContextDone(ctx) {
+				err = ErrOperationTimedOut
+				return
+			}
+
 			if !ok {
-				pair, err = cache.delegate.ResolveKey(keyID)
+				pair, err = cache.delegate.ResolveKey(ctx, keyID)
+				if isContextDone(ctx) {
```

Comment:
You don't need to check the context for done-ness so often.  Also, `ctx.Err() != nil` is the only check you need to determine if the context has been canceled (which includes timeouts).